### PR TITLE
MANTA-4604 Remove electric-boray from the buckets request path

### PR DESCRIPTION
**Pull request by bahamas10**
This is a combination of changes from @kellymclaughlin and me.

* https://github.com/joyent/manta-buckets-api/pull/6
* https://github.com/joyent/node-boray/pull/11
* https://github.com/joyent/electric-boray/pull/9

### CODE REVIEW

File: package.json
Comment (bahamas10) [inline package.json]: This perhaps should change since `node-boray` has its own PR

Comment (kellymclaughlin) [reply package.json]: Let's just wait and merge the `node-boray`/`node-buckets-mdapi` PR and then update this  right before merging.

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js:136]: I'm not sure what this means to assign _pollTimer here but also do it inside pollPlacementData(). Which invocation of setTimeout ends up as _pollTimer in this scenario?

I think we could also have pollPlacementData() as a method on MetadataPlacement and just call that method here instead, but I understand not doing that in order to keep the method private. I think that's just the JS world we live in so happy to see it done another way if preferred.

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: I feel a bit dumb for this one, but I've not seen a require on its own like this. Is there something in "./errors" that we want to execute but not store here?

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: Should this be `_pollTimer`?

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: I think this might need to iterate over `electric_boray_client` and `boray_clients` to close each one.

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: This message doesn't need to include "MetadataPlacement" because the parent's `log` is a child logger with `component` set.

It looks like this applies to most of the log messages here.

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: Rogue indentation.

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: Rogue indentation.

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: I think this is done inside `getBucketLocation()` also. It looks like we only do it twice so far, but it's quite a complicated bit of code to read, so would it be possible to lift it into a utility function?

File: lib/metadata_placement.js
Comment (chudley) [inline lib/metadata_placement.js]: `moray` here is probably going to change with the renaming soon, but thought I'd highlight its use here just in case.

File: lib/boray_client.js
Comment (chudley) [inline lib/boray_client.js]: I don't think we have a standard for this, but I've mostly seen child loggers set their component to upper camel case.

I think I'd also prefer to see the pnode as its own property on this child logger instead of a suffix of the component name. Not a big deal and I'm not sure whether it'll realistically be used, but it will perhaps bring more flexibility to log searching in the future.

File: lib/boray_client.js
Comment (chudley) [inline lib/boray_client.js:52]: I was going to mention an assertion after the `parseInt` usage, but I would expect cueball to also do this so this is fine as it is.

File: lib/buckets/common.js
Comment (chudley) [inline lib/buckets/common.js]: I think I would prefer `requestId` (or `request_id`) for this variable name. It's a little ambiguous given the other "*id"s in this function, and it's also not used until a fair bit later on.

File: lib/common.js
Comment (chudley) [inline lib/common.js]: Looks like these could possibly be removed.

File: lib/buckets/common.js
Comment (chudley) [inline lib/buckets/common.js:299]: This looks different to the node-lomstream library we have, but do they serve the same purpose?

**Review by chudley — COMMENTED** [review]
This change is quite big, so hopefully you don't mind me taking a few passes over it!

Some of this work around bucket/object listings (such as `_list` and `LimitMarkerStream`) I think might be lifted from another project into this one, so I haven't reviewed it too heavily as it's likely already received some scrutiny. I would be happy to approve those bits as they are if you're confident that they've already been reviewed as part of integration to the other project.

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js:136]: I have to be honest and say I copied the `picker` module and modified it to suit the needs for metadata polling so I don't know the *why* to all the bits of code here other than "it's what picker does". We're not bound to it so if you think having a method is a better way then that's ok by me.

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js]: More copypasta that can probably just be removed in this case.

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js]: yes, nice catch!

File: lib/buckets/common.js
Comment (kellymclaughlin) [reply lib/buckets/common.js]: Yep, will change

File: lib/metadata_placement.js
Comment (bahamas10) [reply lib/metadata_placement.js]: `./errors.js` does overload `global` with a set of error messages.  It's not at all my favorite pattern, and should probably be refactored to just return the errors as opposed to attaching them to global.

A quick look through this file though shows that none of those exposed errors from `./errors.js` are used here.

File: lib/buckets/common.js
Comment (bahamas10) [reply lib/buckets/common.js:299]: So, this was originally a LimitOffset (no marker) stream before the listing API was updated a couple of months ago.  I had already written code for that, so I just adapted it to work with a Limit and a Marker.

It probably could be rewritten to use node-lomstream at this point (though I have little experience with that library).  I don't know if they do exactly the same thing however.

**Comment by bahamas10** [comment]
@chudley you are exactly right about `_list` and `LimitMarkerStream` having already been reviewed before they were merged into electric-boray.  The only thing I have added in this review of note is the block comments above `paginationStream` and `LimitMarkerStream` - so I would happily welcome any review on that :)

File: package.json
Comment (kellymclaughlin) [reply package.json]: Addressed, it's pointing to `master` now which is itself also a temporary condition.

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js]: Removed

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js]: Fixed

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js]: Done

File: lib/metadata_placement.js
Comment (kellymclaughlin) [reply lib/metadata_placement.js]: Thanks for pointing that out. I've fixed those up.

File: lib/boray_client.js
Comment (kellymclaughlin) [reply lib/boray_client.js]: Done

File: lib/buckets/common.js
Comment (kellymclaughlin) [reply lib/buckets/common.js]: Changed

File: lib/common.js
Comment (kellymclaughlin) [reply lib/common.js]: I removed those two lines.

File: lib/metadata_placement.js
Comment (bahamas10) [reply lib/metadata_placement.js]: agreed, will do.

File: lib/metadata_placement.js
Comment (bahamas10) [reply lib/metadata_placement.js]: will do

File: lib/metadata_placement.js
Comment (bahamas10) [reply lib/metadata_placement.js]: will do

File: lib/buckets/common.js
Comment (chudley) [reply lib/buckets/common.js:299]: > I don't know if they do exactly the same thing however.

Yea me neither off hand. Seems fine to keep muskie/buckets-api's own implementation if it's already working!

File: lib/metadata_placement.js
Comment (chudley) [reply lib/metadata_placement.js:136]: I think it's probably fine for now given what this change is intended for. As we get more sensitive to data placement changes then perhaps we can revisit to make sure the timing is doing what we expect, but if it's working for picker then I'm happy!

File: lib/buckets/common.js
Comment (chudley) [inline lib/buckets/common.js:19]: Sorry, just noticed this funky import as well.

**Review by chudley — DISMISSED** [review]
Thanks @bahamas10. I checked out the comments you added and they look good.

And thanks both for addressing my feedback! I've had another look over the change as a whole and it looks good to me.

I only have one minor comment that I've added on this latest round, but for the sake of quick iteration I'm happy for it to be addressed in a later change. It won't affect functionality of this service; it's just a little unsightly.

File: lib/buckets/common.js
Comment (bahamas10) [reply lib/buckets/common.js:19]: This is used to get `InvalidLimitError` defined in this file.  I'll ask about this in ~manta, but I think it would be best to revamp `errors.js` to:

1. not abuse `global`, and
2. not `fs.readFileSync` itself to create exports

as part of a separate PR

Comment (kellymclaughlin) [reply lib/buckets/common.js:19]: +1 to that idea, @bahamas10

**Review by chudley — APPROVED** [review]

**Pull request merged**
